### DISCUSSION TITLE
Questions about Bluetooth streaming

### DISCUSSION THREAD

**Discussion by craiglindley** [pschatzmann/arduino-audio-tools#26] [Q&A] [answered]
I am investigating the idea of streaming Internet radio to a bluetooth speaker. I understand that the esp32 cannot do both WiFi and Bluetooth at the same time so I was thinking of using 2 esp32's and connecting them together using i2s. Which brings me to my question. Does the Bluetooth streamer only run at 44100, 16 bit? I am asking because various Internet radio stations stream at 320kbps or 128k bps so I am not sure how to couple the url streamer on one esp32 to the Bluetooth streamer (a2dp) on another esp32. Any help would be appreciated.

Nice library BTW.

**Comment by pschatzmann** [comment, marked as answer]
Interesting question. One small disclaimer: I did't try this myself so I don't know all the stumbling blocks!
I think most likely you will be streaming mp3. So the relevant information is provided by the codec. I noticed that most often I am getting the following output:
- sample_rate: 44100
- bits_per_sample: 16
- channels: 2
And this is exactly what you need for A2DP as well!

I suggest that you experiment first a little bit with the following sketch:
https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-player/player-url-i2s

So there is a good chance that in most cases it might just work...

**Comment by pschatzmann** [comment]
What I forgot to mention: As far as I know A2DP supports some adaptive speed adjustments and can automatically switch between the following rates
- 16000
- 32000
- 44100
- 48000

**Comment by craiglindley** [comment]
Excellent thanks for the quick response.  will try some things and report
back.

On Wed, Oct 27, 2021 at 11:39 AM Phil Schatzmann ***@***.***>
wrote:

> What I forgot to mention: As far as I know A2DP supports some adapting
> speed adjustments and can automatically switch between the following rates
>
>    - 16000
>    - 32000
>    - 44100
>    - 48000
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pschatzmann/arduino-audio-tools/discussions/26#discussioncomment-1547624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH6K5D2I4G5IW2GFR7NJWQ3UJBBN7ANCNFSM5G23DGDQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Craig Lindley / Heather Hubbard

New Recordings are here <http://craigandheather.net/spellbound.html>

Personal Website is here <http://craigandheather.net>

Please call the Rockrimmon house. Our cell phones don't work at home.
Rockrimmon House: (719) 426-9873
Craig's Cell: (719) 502-7925
Heather's Cell: (719) 571-0944

If you’re one in a million, there are now more than seven thousand people
exactly like you.

**Reply by craiglindley** [reply]
So now I have a question about the streams-i2s-a2dp.ino sketch. I got it to compile successfully and to download into my esp32. I assume that it should connect to my BT amplifier using the name of MyMusic but it does not.  It stalls at the a2dpStream.begin(TX_MODE, "MyMusic");  function and does nothing else. Any ideas of what I am doing wrong?

**Reply by craiglindley** [reply]
Never mind I got it to work

**Reply by pschatzmann** [reply]
What was the problem?

**Comment by craiglindley** [comment]
My stupidity. I left out a letter in the bluetooth name.
But now I have a different problem which I am researching. I connected up
two esp32s using i2s as follows:

ESP32 #1           ESP32 #2
      Vin                     Vin
      Gnd                    Gnd
      22        SD          32
      21        WS         15
         5       SCK        14

Where ESP32 #2 is the device which transmits the Bluetooth audio.
I first ran your sketch streams-generator-a2dp.ino on esp32 #2 and got a
nice tone from my speaker. I then replaced that sketch
with streams-i2s-a2dp.ino and put streams-generator-i2s.ino on esp32 #1
after changing the pin numbers of course and got nothing but random garbage
out. Since I wired Vin/Gnd in both devices together they boot at the same
time and I wonder if this is part of the problem.

I'll let you know what I find out.





On Thu, Oct 28, 2021 at 10:53 PM Phil Schatzmann ***@***.***>
wrote:

> What was the problem?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pschatzmann/arduino-audio-tools/discussions/26#discussioncomment-1556758>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AH6K5D3ILQMM26MFITAQ4TTUJIZGXANCNFSM5G23DGDQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Craig Lindley / Heather Hubbard

New Recordings are here <http://craigandheather.net/spellbound.html>

Personal Website is here <http://craigandheather.net>

Please call the Rockrimmon house. Our cell phones don't work at home.
Rockrimmon House: (719) 426-9873
Craig's Cell: (719) 502-7925
Heather's Cell: (719) 571-0944

If you’re one in a million, there are now more than seven thousand people
exactly like you.

**Reply by pschatzmann** [reply]
Don't forget to define one device as client and the other as master with the help of config.is_master  ...

**Reply by craiglindley** [reply]
I tried that this afternoon and I also changed the pin connections for esp32#1 (SD=22, WS=15, SCK=14) and neither had a positive affect.

**Reply by craiglindley** [reply]
More info:

I tried making esp32 #1 master and esp32 #2 slave with no effect. I then reversed this and made esp32 #1 slave and esp32 #2 master with no effect. The noise I am hearing from my BT speaker is however definitely related to the note specified in the sine wave.begin function. If I change the note to B5 the pitch of the noise in the speaker goes up accordingly. So it seems like data is being moved between the two devices but is either corrupted or it is underflowing or overflowing the receiving end.

Thoughts?

**Reply by craiglindley** [reply]
In case anyone else would like to experiment as well the current wiring is as follows:

ESP32 #1           ESP32 #2
      Vin                     Vin
      Gnd                    Gnd
      22        SD          32
      15        WS         15
       14       SCK        14

with esp32#1 being the tone generator and esp32#2 being the BT transmitter.

Code for esp32#1 is:

`#include "AudioTools.h"

using namespace audio_tools;

#define SAMPLE_RATE  44100
#define CHANNELS         2
#define BITS_PER_SAMPLE 16

typedef int16_t sound_t;                                   // sound will be represented as int16_t (with 2 bytes)
SineWaveGenerator<sound_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<sound_t> sound(sineWave);             // Stream generated from sine wave
I2SStream i2sStream;
StreamCopy copier(i2sStream, sound);                       // copies sound into i2s

// Arduino Setup
void setup(void) {

  Serial.begin(115200);
  delay(1000);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // start i2s output
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(TX_MODE);
  config.is_master = false;
  config.pin_data = 22;
  config.pin_ws = 15;
  config.pin_bck = 14;
  config.sample_rate = SAMPLE_RATE;
  config.channels = CHANNELS;
  config.bits_per_sample = BITS_PER_SAMPLE;
  i2sStream.begin(config);

  // Setup sine wave
  sineWave.begin(CHANNELS, SAMPLE_RATE, N_B4);
}

// Arduino loop - copy sound to i2sStream
void loop() {
  copier.copy();
}`

and code for esp32#2 is:

`// Add this in your sketch or change the setting in AudioConfig.h
#define USE_A2DP

#include "Arduino.h"
#include "AudioTools.h"
#include "AudioA2DP.h"

using namespace audio_tools;

#define SAMPLE_RATE  44100
#define CHANNELS         2
#define BITS_PER_SAMPLE 16

I2SStream i2sStream;                        // Access I2S as stream
A2DPStream a2dpStream = A2DPStream::instance(); // access A2DP as stream
StreamCopy copier(a2dpStream, i2sStream); // copy i2sStream to a2dpStream

// Arduino Setup
void setup(void) {

  Serial.begin(115200);
  delay(1000);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // start bluetooth
  Serial.println("starting A2DP...");
  a2dpStream.begin(TX_MODE, "KD-240WF");
  Serial.println("A2DP is connected now");

  // start i2s input
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(RX_MODE);
  config.is_master = true;
  config.pin_data = 32;
  config.pin_ws = 15;
  config.pin_bck = 14;
  config.sample_rate = SAMPLE_RATE;
  config.channels = CHANNELS;
  config.bits_per_sample = BITS_PER_SAMPLE;
  i2sStream.begin(config);
  Serial.println("I2S started");
}

// Arduino loop - copy i2s data to bluetooth
void loop() {
  if (a2dpStream) {
    copier.copy();
  }
}`

If anyone sees any errors in this code please let me know. At this point I am kind of stumped as to what is going on.

**Reply by pschatzmann** [reply]
My suggestion would be to reduce the complexity and first just try to setup and test the I2S connection (w/o A2DP). 
You could try to use the CSVStream class in the second sketch and display the data in the Serial Plotter just to double check that we receive a clean audio signal.

Another parameter that you could play with is the config. i2s_mode (with the values I2S_LSB_MODE , I2S_MSB_MODE) - though I would be surprised that this is the source of the problem.

**Reply by craiglindley** [reply]
Excellent suggestion which I did. The resultant serial plotter image is attached:

![Screen Shot 2021-10-30 at 2 48 50 PM](https://user-images.githubusercontent.com/33336975/139558668-18eb06ad-5de5-4b29-9268-5881272bb880.png)

As you can see the sine wave is getting through but it is being cut off periodically and that is the reason for the noise I am hearing through my BT speaker.

I then changed the channel count to 1 and the sine wave comes through clearly although I do hear some crackles which might be due to only a single channel being fed to the BT speaker.

Any ideas on why the stream is being interrupted?

**Reply by craiglindley** [reply]
I now know why. The AudioLogger. If I comment out the following line in the code for esp32#1

 // AudioLogger::instance().begin(Serial, AudioLogger::Info);

the major dropouts go away completely. There is still however, crackling in the BT output but that might be discontinuities in the sine wave generation. I'll next run audio from spiffs to see if it is impacted

**Reply by pschatzmann** [reply]
Good point - I was usually using lower sample rates...
You could alternatively use 
- AudioLogger::instance().begin(Serial, AudioLogger::Error);
- AudioLogger::instance().begin(Serial, AudioLogger::Warning);
to request another log level. The default is Warning

**Reply by pschatzmann** [reply]
Today I was trying to do some improvements to the A2DPStream class in relation to the AudioPlayer and finally have the MP3 Decoding from a SD Drive with output to A2DP working with a perfect quality. 

Maybe this also helps with your quality issues.

**Reply by monojk** [reply]
> I now know why. The AudioLogger. If I comment out the following line in the code for esp32#1
> 
> // AudioLogger::instance().begin(Serial, AudioLogger::Info);
> 
> the major dropouts go away completely. There is still however, crackling in the BT output but that might be discontinuities in the sine wave generation. I'll next run audio from spiffs to see if it is impacted

Did you manage to get a good quality. I am currently struggling with distortions: https://github.com/pschatzmann/arduino-audio-tools/discussions/149

**Comment by podaen** [comment]
If you have the posibility to add a button to switch between a2dp and internet radio... that would be the most easiest way there is.

**Reply by pschatzmann** [reply]
Since the releasing of the memory (and restarting of A2DP) is not working reliably I would store the requested scenario in the NVS and reboot. In the setup, the start up logic would consult the NVS to select either to start up the A2DP or the WIFI.

**Comment by BlackBirdDevelopments** [comment]
I would like to join this discussion, but with a different question.
@pschatzmann as you already know, I got a stream from i2s to a2dp. Everything works fine and I have no issues.
In addition, I would like to visualize the sound I hear or the sound from the mic. As far as I know, the microphone gives me digital data and the data sent to the BT speaker are as well digital. Is there a point in the code where I can cut in and either print the data as analog to a file (while connected to the PC) or assign it to a pin, where I can visualize it than with an oscilloscope? 

Another thing, It might be nice to have the option of connecting to my earbuds (that already works perfectly) or to connect to my phone and then use recording software to either visualize or just for showing purpose. Issues could be, that the phone should be the Master in this case and to change the BT identifier of the ESP, in a manner that everything works together.

**Reply by pschatzmann** [reply]
If I understand you want to  have access to the processed audio data. I can give you 4 different approaches:
1) Currently you use my copy helper class, but you could just replace it with a readBytes to get the data and do a write(data, len) to the A2DPStream and your additional processing.
2) I have a callback method (setCallbackOnWrite) in the StreamCopy class which you could use. It would call the callback that you have provided and provides the actually processed audio data. I am currently using this mechanism to extract e.g. the metadata.
3) You could create your own special purpose output class (see MetaDataPrint) and in your sketch wrap it with the A2DPStream in a MultiOutput class which will be the copy target. This way you end up with a reusable visualization class that gets the data along with the A2DPStream.
4) You could create a subclass of the A2DPStream class and overwrite the write (const uint8_t *data, size_t len) method. You would add you own logic and also call A2DPStream::write(data,len) from there. This way you get a reusable A2DP visualization class.

Just pick the one that makes most sense for your project...



<br class="Apple-interchange-newline">
